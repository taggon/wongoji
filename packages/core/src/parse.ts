import { tokenize, type Token } from './tokenize';
import type { WongojiConfig } from './config';

export type Document = {
  type: 'document';
  nodes: (Paragraph | Quote | SingleQuote)[];
};

export type Paragraph = {
  type: 'paragraph';
  nodes: (FullText | HalfText | Space | Punctuation | Quote | SingleQuote)[];
};

export type Quote = {
  type: 'quote';
  nodes: (FullText | HalfText | Space | Punctuation | SingleQuote)[];
  inline: boolean;
};

export type SingleQuote = {
  type: 'singlequote';
  nodes: (FullText | HalfText | Space | Punctuation)[];
  inline: boolean;
};

export type Space = {
  type: 'space';
};

export type FullText = {
  type: 'fulltext';
  raw: string;
};

export type HalfText = {
  type: 'halftext';
  raw: string;
};

export type Punctuation = {
  type: 'punctuation';
  raw: string;
};

export type ParentNode = Document | Paragraph | Quote | SingleQuote;

export type Node = ParentNode | Space | FullText | HalfText | Punctuation;

function eat(tokens: Token[], types: Token['type'][], start: number) {
  let i = start - 1;

  while (types.includes(tokens[++i]?.type));

  return i - 1;
}

function parseTokensIntoNodes(parent: ParentNode, tokens: Token[], from: number, to: number) {
  let i = from - 1;

  while(++i < tokens.length && i < to) {
    const tok = tokens[i];

    switch (tok.type) {
      case 'fulltext':
      case 'halftext':
      case 'punctuation':
        // 새 문단을 시작한다.
        if (parent.type === 'document') {
          const start = i;
          const endOfParagraph = tokens.findIndex((t, idx) => {
            return idx > i && t.type === 'newline';
          });
          i = endOfParagraph === -1 ? tokens.length : endOfParagraph;

          const paragraph: Paragraph = { type: 'paragraph', nodes: [] };
          parseTokensIntoNodes(paragraph, tokens, start, i);
          parent.nodes.push(paragraph);
          break;
        }

        parent.nodes.push({
          type: tok.type,
          raw: tok.raw,
        });

        // 문장 부호라면 뒤따른 공백을 무시한 다음...
        if (tok.type === 'punctuation') {
          i = eat(tokens, ['space'], i + 1);
        }
        break;
      case 'space':
        // 최상위 노드에는 공백이 허용되지 않는다.
        if (parent.type === 'document') continue;
        if (parent.nodes.length > 0) {
          // 바로 앞 노드가 인용문이라면 공백을 무시한다.
          const lastNode = parent.nodes[parent.nodes.length - 1];
          if (lastNode.type === 'quote' || lastNode.type === 'singlequote') {
            continue;
          }
          parent.nodes.push({ type: 'space' });
        }
        break;
      case 'newline':
        // 최상위 노드에는 줄바꿈이 허용되지 않는다.
        if (parent.type === 'document') continue;
        if (parent.nodes.length > 0) {
          parent.nodes.push({ type: 'space' });
        }
        break;
      case 'quote':
        // 닫는 따옴표의 인덱스를 찾는다.
        const start = i + 1;
        const closingQuoteIndex = tokens.findIndex((t, idx) => {
          return idx > i && t.type === 'quote' && t.raw === tok.raw;
        });
        i = closingQuoteIndex;

        if (closingQuoteIndex === -1) {
          // TODO: 닫는 따옴표를 찾지 못했을 때의 에러 처리
        } else {
          const quote: Quote | SingleQuote = { type: tok.raw === '"' ? 'quote' :  'singlequote', nodes: [], inline: parent.type != 'document' };
          parseTokensIntoNodes(quote, tokens, start, closingQuoteIndex);

          // 문단을 따옴표로 시작했지만, 보통 문단의 일부인 경우
          if (parent.type === 'document') {
            i = eat(tokens, ['space'], i + 1);
            if (['fulltext', 'halftext'].includes(tokens[i + 1]?.type)) {
              const paragraph: Paragraph = { type: 'paragraph', nodes: [quote] };
              parent.nodes.push(paragraph);

              // 뒤이어 나오는 문장을 같은 문단으로 처리한다.
              parent = paragraph;
              quote.inline = true;
              continue;
            }
          }

          if (parent.type === 'quote' || parent.type === 'singlequote') {
            // 따옴표가 중첩되는 경우가 아니면 여기를 탈 일이 없을텐데...
            break;
          }

          parent.nodes.push(quote);
        }
        break;
    }
  }
}

export function parse(text: string, config: WongojiConfig = {}): Document {
  if (config.replaceDotsWithEllipsis) {
    text = text.replace(/\.{3}/g, '…');
  }

  const tokens = tokenize(text);
  const doc: Document = { type: 'document', nodes: [] };

  parseTokensIntoNodes(doc, tokens, 0, tokens.length);

  return doc;
}
