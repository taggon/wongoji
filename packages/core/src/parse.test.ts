import { expect, test } from 'bun:test';
import { parse } from './parse';

const testCases = [
  '안녕하세요. 반갑습니다.',
  '대한민국은 Korea이다.',
  '걸음도 해깝고 방울소리가 밤 벌판에 한층 청청하게 울렸다.\n달이 어지간히 기울어졌다.',
  '"이걸 까셀라 부다!"하고 소리를 쳤다.',
  '대체 이게 웬 속인지(지금까지도 난 영문을 모른다) 아버질 혼내 주기는 제가 내래 놓고 이제 와서는 달려들며, "에그머니! 이 망할 게 아버지 죽이네!" 하고 내 귀를 뒤로 잡아당기며 마냥 우는 것이 아니냐.'
];

testCases.forEach((text, index) => {
  test(`Case #${index + 1}: ${text}`, () => {
    expect(parse(text)).toMatchSnapshot();
  });
});

test('replaceDotsWithEllipsis 옵션을 켜면 점 세 개를 …으로 변경한다.', () => {
  expect(parse('그런...', { replaceDotsWithEllipsis: false })).toMatchSnapshot();
  expect(parse('그런...', { replaceDotsWithEllipsis: true })).toMatchSnapshot();
});
