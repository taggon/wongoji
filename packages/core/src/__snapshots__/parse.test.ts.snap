// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Case #1: 안녕하세요. 반갑습니다. 1`] = `
{
  "nodes": [
    {
      "nodes": [
        {
          "raw": "안녕하세요",
          "type": "fulltext",
        },
        {
          "raw": ".",
          "type": "punctuation",
        },
        {
          "raw": "반갑습니다",
          "type": "fulltext",
        },
        {
          "raw": ".",
          "type": "punctuation",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`Case #2: 대한민국은 Korea이다. 1`] = `
{
  "nodes": [
    {
      "nodes": [
        {
          "raw": "대한민국은",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "K",
          "type": "fulltext",
        },
        {
          "raw": "orea",
          "type": "halftext",
        },
        {
          "raw": "이다",
          "type": "fulltext",
        },
        {
          "raw": ".",
          "type": "punctuation",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`Case #3: 걸음도 해깝고 방울소리가 밤 벌판에 한층 청청하게 울렸다.
달이 어지간히 기울어졌다. 1`] = `
{
  "nodes": [
    {
      "nodes": [
        {
          "raw": "걸음도",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "해깝고",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "방울소리가",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "밤",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "벌판에",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "한층",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "청청하게",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "울렸다",
          "type": "fulltext",
        },
        {
          "raw": ".",
          "type": "punctuation",
        },
      ],
      "type": "paragraph",
    },
    {
      "nodes": [
        {
          "raw": "달이",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "어지간히",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "기울어졌다",
          "type": "fulltext",
        },
        {
          "raw": ".",
          "type": "punctuation",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`Case #4: "이걸 까셀라 부다!"하고 소리를 쳤다. 1`] = `
{
  "nodes": [
    {
      "nodes": [
        {
          "inline": true,
          "nodes": [
            {
              "raw": "이걸",
              "type": "fulltext",
            },
            {
              "type": "space",
            },
            {
              "raw": "까셀라",
              "type": "fulltext",
            },
            {
              "type": "space",
            },
            {
              "raw": "부다",
              "type": "fulltext",
            },
            {
              "raw": "!",
              "type": "punctuation",
            },
          ],
          "type": "quote",
        },
        {
          "raw": "하고",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "소리를",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "쳤다",
          "type": "fulltext",
        },
        {
          "raw": ".",
          "type": "punctuation",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`Case #5: 대체 이게 웬 속인지(지금까지도 난 영문을 모른다) 아버질 혼내 주기는 제가 내래 놓고 이제 와서는 달려들며, "에그머니! 이 망할 게 아버지 죽이네!" 하고 내 귀를 뒤로 잡아당기며 마냥 우는 것이 아니냐. 1`] = `
{
  "nodes": [
    {
      "nodes": [
        {
          "raw": "대체",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "이게",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "웬",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "속인지(지금까지도",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "난",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "영문을",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "모른다)",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "아버질",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "혼내",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "주기는",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "제가",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "내래",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "놓고",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "이제",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "와서는",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "달려들며",
          "type": "fulltext",
        },
        {
          "raw": ",",
          "type": "punctuation",
        },
        {
          "inline": true,
          "nodes": [
            {
              "raw": "에그머니",
              "type": "fulltext",
            },
            {
              "raw": "!",
              "type": "punctuation",
            },
            {
              "raw": "이",
              "type": "fulltext",
            },
            {
              "type": "space",
            },
            {
              "raw": "망할",
              "type": "fulltext",
            },
            {
              "type": "space",
            },
            {
              "raw": "게",
              "type": "fulltext",
            },
            {
              "type": "space",
            },
            {
              "raw": "아버지",
              "type": "fulltext",
            },
            {
              "type": "space",
            },
            {
              "raw": "죽이네",
              "type": "fulltext",
            },
            {
              "raw": "!",
              "type": "punctuation",
            },
          ],
          "type": "quote",
        },
        {
          "raw": "하고",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "내",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "귀를",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "뒤로",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "잡아당기며",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "마냥",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "우는",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "것이",
          "type": "fulltext",
        },
        {
          "type": "space",
        },
        {
          "raw": "아니냐",
          "type": "fulltext",
        },
        {
          "raw": ".",
          "type": "punctuation",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`replaceDotsWithEllipsis 옵션을 켜면 점 세 개를 …으로 변경한다. 1`] = `
{
  "nodes": [
    {
      "nodes": [
        {
          "raw": "그런",
          "type": "fulltext",
        },
        {
          "raw": ".",
          "type": "punctuation",
        },
        {
          "raw": ".",
          "type": "punctuation",
        },
        {
          "raw": ".",
          "type": "punctuation",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`replaceDotsWithEllipsis 옵션을 켜면 점 세 개를 …으로 변경한다. 2`] = `
{
  "nodes": [
    {
      "nodes": [
        {
          "raw": "그런…",
          "type": "fulltext",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;
